dart/lib/src/recursive_exception_cause_extractor.dart

```diff
@@ -0,0 +1,47 @@
+import 'sentry_options.dart';
+import 'throwable_mechanism.dart';
+import 'exception_cause.dart';
+import 'protocol.dart';
+
+/// Extracts inner exceptions recursively
+class RecursiveExceptionCauseExtractor {
+  RecursiveExceptionCauseExtractor(this._options);
+
+  final SentryOptions _options;
+
+  List<ExceptionCause> flatten(exception, stackTrace) {
+    final allExceptionCauses = <ExceptionCause>[];
+    final circularityDetector = <dynamic>{};
+
+    var currentException = exception;
+    ExceptionCause? currentExceptionCause =
+        ExceptionCause(exception, stackTrace);
+
+    while (currentException != null &&
+        currentExceptionCause != null &&
+        circularityDetector.add(currentException)) {
+      allExceptionCauses.add(currentExceptionCause);
+
+      final extractionSourceSource = currentException is ThrowableMechanism
+          ? currentException.throwable
+          : currentException;
+
+      final extractor =
+          _options.exceptionCauseExtractor(extractionSourceSource.runtimeType);
+
+      try {
+        currentExceptionCause = extractor?.cause(extractionSourceSource);
+        currentException = currentExceptionCause?.exception;
+      } catch (exception, stackTrace) {
+        _options.logger(
+          SentryLevel.error,
+          'An exception occurred while extracting  exception cause',
+          exception: exception,
+          stackTrace: stackTrace,
+        );
+        break;
```

Comment:
I guess we can skip the `break` here and let it go with the other items of the list.

Reply:
We should still report the root issue even if the cause is a problem, the root is an exception by itself.

Reply:
Since it is assigned before the exception happens, it is part of the returned list. Its in the throwing test that i added `expect(actual, [errorA]);` Think this is covered, or do you see a potential issue?

Reply:
Indeed, missed this test, thanks.